Allow dispatch based on ScalarType, Layout, Device

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28494 [wip] Implement backend level fallback for c10
* #28314 Simplify c10 dispatcher
* #28251 Move unboxed dispatch decision into dispatcher
* **#28723 Allow dispatch based on ScalarType, Layout, Device**

Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

smessmer committed:
Allow dispatch based on ScalarType, Layout, Device

Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

[ghstack-poisoned]

smessmer committed:
Update on "Allow dispatch based on ScalarType, Layout, Device"


Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

[ghstack-poisoned]

smessmer committed:
Update on "Allow dispatch based on ScalarType, Layout, Device"


Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -98,7 +98,10 @@ namespace detail {
+      TensorOptions tensorOptions = TensorOptionsAccumulator().apply(args...).options;
+      if (tensorOptions.has_dtype() && tensorOptions.has_device() && tensorOptions.has_layout()) {
+        type_set(tensorOptions);
+      }
```
Yeah, so replicating the comment I made earlier, I think that we should probably just make TensorOptions return a type set rather than a type id, and then just some configurations of options just result in an empty set being returned. (Or I guess it could return `optional<TypeId>`, I can't think of any configurations where we'd return TWO type ids.)

smessmer committed:
Update on "Allow dispatch based on ScalarType, Layout, Device"


Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

[ghstack-poisoned]

smessmer committed:
Update on "Allow dispatch based on ScalarType, Layout, Device"


Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 86c1add:
* 1/2 flaky
* 1/2 broken upstream



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=86c1addd428aaaac3a8a58381c43ac69795453fe).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

smessmer committed:
Update on "Allow dispatch based on ScalarType, Layout, Device"


Dispatch based on TensorOptions was already a thing, but in the c10 dispatcher, these are unpacked into ScalarType, Layout and Device.
So to correctly treat dispatch for these operators in c10, we need to look for the unpacked types.

Differential Revision: [D18151211](https://our.internmc.facebook.com/intern/diff/D18151211/)

[ghstack-poisoned]
